### PR TITLE
Add main async code match test

### PR DESCRIPTION
**Pull request by mayureshagashe2105**
## Description of the change
Add a centralized **async code match test**

Might need some tweaking here. This is a very raw implementation.

## Motivation
Remove repetition for each module

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added detailed comments to my code where applicable.
- [x] I have verified that my change does not break existing code.
- [x] My PR is based on the latest changes of the main branch (if unsure, please run `git pull --rebase upstream main`).
- [x] I am familiar with the [Google Style Guide](https://google.github.io/styleguide/) for the language I have coded in.
- [x] I have read through the [Contributing Guide](https://github.com/google/generative-ai-python/blob/main/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

File: google/generativeai/answer.py [hunk 1]
```diff
@@ -354,6 +358,6 @@ async def generate_answer_async(
         answer_style=answer_style,
     )
 
-    response = await client.generate_answer(request)
+    response = await client.generate_answer(request, **request_options)
```
Comment (MarkDaoust) [inline google/generativeai/answer.py:361]: I guess you found this because the test works? LGTM.

File: tests/test_async_code_match.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            .replace("async ", "")
+            .replace("await ", "")
+            .replace("Async", "")
+            .replace("ASYNC_", "")
+        )
+        asource = re.sub(" *?# type: ignore", "", asource)
+        self.assertEqual(source, asource)
+
+    def test_code_match_for_async_methods(self):
+        for working_dir in WORKING_DIRS:
```
Comment (MarkDaoust) [inline tests/test_async_code_match.py:59]: Nit: only one directory, maybe drop.

File: tests/test_async_code_match.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+WORKING_DIRS = [
+    (pathlib.Path(__file__).parent.parent / "google"),
+]
+EXEMPT_DIRS = ["notebook"]
+EXEMPT_DECORATORS = ["overload", "property", "setter", "abstractmethod", "staticmethod"]
+EXEMPT_FILES = ["client.py", "version.py", "discuss.py", "files.py"]
```
Comment (MarkDaoust) [inline tests/test_async_code_match.py:29]: Does the test still pass if these lists are empty?

File: tests/test_async_code_match.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            .replace("Async", "")
+            .replace("ASYNC_", "")
+        )
+        asource = re.sub(" *?# type: ignore", "", asource)
+        self.assertEqual(source, asource)
+
+    def test_code_match_for_async_methods(self):
+        for working_dir in WORKING_DIRS:
+            for fpath in working_dir.rglob("*.py"):
+                if fpath.name.split("/")[-1] in EXEMPT_FILES or any(
```
Comment (MarkDaoust) [inline tests/test_async_code_match.py:61]: Nit isn't `fpath.name` == `fpath.split("/")[-1]`?

File: tests/test_async_code_match.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+
+                for node in ast.walk(source_nodes):
+                    if isinstance(
+                        node, (ast.FunctionDef, ast.AsyncFunctionDef)
+                    ) and not node.name.startswith("_"):
+                        name = node.name[:-6] if node.name.endswith("_async") else node.name
+                        if name in EXEMPT_FUNCTIONS:
+                            continue
+                        # print(f"Checking {node.name}")
+                        is_exempt = False
```
Comment (MarkDaoust) [inline tests/test_async_code_match.py:78]: Maybe pack this decorator-bloc into a stand alone function.

**Review by MarkDaoust — COMMENTED** [review]
How does this handle GenerateContentResponse vs. AsyncGenerateContentResponse?

https://github.com/google-gemini/generative-ai-python/blob/37eb4cb/google/generativeai/types/generation_types.py#L446-L576

File: google/generativeai/answer.py (hunk 1 above)
Comment (mayureshagashe2105) [reply google/generativeai/answer.py:361]: yes! :)

File: tests/test_async_code_match.py (hunk 3 above)
Comment (mayureshagashe2105) [reply tests/test_async_code_match.py:29]: no!

So for files.py, we don't support async implementation, so it fails there.
For discuss.py, there are many inconsistencies with error messages and decorators, (so basically test fails as the code is not exactly the same/consistent)

**Comment by mayureshagashe2105** [comment]
> How does this handle GenerateContentResponse vs. AsyncGenerateContentResponse?
> 
> https://github.com/google-gemini/generative-ai-python/blob/37eb4cb/google/generativeai/types/generation_types.py#L446-L576

It doesn't. It checks individual methods though.

File: tests/test_async_code_match.py (hunk 4 above)
Comment (mayureshagashe2105) [reply tests/test_async_code_match.py:61]: Done.

File: tests/test_async_code_match.py (hunk 5 above)
Comment (mayureshagashe2105) [reply tests/test_async_code_match.py:78]: Done. Thanks!

**Review by MarkDaoust — APPROVED** [review]

**Pull request merged**